### PR TITLE
[Refactor] Refactor images pull command

### PR DESCRIPTION
**Pull request by suyanhanx**
part of #1680 

Checklist:

* [x]  Create a file in `pkg/api/types/${cmd}_types.go`, and define the CommandOption for this command
* [x]  Create some file in `pkg/cmd/${cmd}`, and move the command entry point in real into this package

Signed-off-by: suyanhanx <suyanhanx@gmail.com>

### CODE REVIEW

File: cmd/nerdctl/pull.go [hunk 1]
```diff
@@ -74,44 +63,55 @@ func newPullCommand() *cobra.Command {
...
+}
+
+func pullAction(cmd *cobra.Command, args []string) error {
+	var pullOptions, err = processPullCommandFlags(cmd)
 	if err != nil {
 		return err
 	}
 
-	_, err = ensureImage(ctx, cmd, client, globalOptions, rawRef, ocispecPlatforms, "always", unpack, quiet)
+	err = image.Pull(cmd.Context(), cmd.OutOrStdout(), cmd.ErrOrStderr(), args[0], pullOptions)
```
Comment (Zheaoli) [inline cmd/nerdctl/pull.go:114]:
```suggestion
	return image.Pull(cmd.Context(), cmd.OutOrStdout(), cmd.ErrOrStderr(), args[0], pullOptions)
```

File: cmd/nerdctl/pull.go [hunk 2]
```diff
@@ -120,73 +120,27 @@ func pullAction(cmd *cobra.Command, args []string) error {
...
-			return nil, err
-		}
-	case "none":
-		logrus.Debugf("verification process skipped")
-	default:
-		return nil, fmt.Errorf("no verifier found: %s", verifier)
+	options := types.PullCommandOptions{
+		GOptions:    globalOptions,
+		Verify:      verifier,
+		IPFSAddress: ipfsAddressStr,
```
Comment (Zheaoli) [inline cmd/nerdctl/pull.go:137]: All command options construct processes should be in processPullCommandFlags

File: cmd/nerdctl/pull.go [hunk 3]
```diff
@@ -120,73 +120,27 @@ func pullAction(cmd *cobra.Command, args []string) error {
 }
 
 func ensureImage(ctx context.Context, cmd *cobra.Command, client *containerd.Client, globalOptions types.GlobalCommandOptions, rawRef string, ocispecPlatforms []v1.Platform, pull string, unpack *bool, quiet bool) (*imgutil.EnsuredImage, error) {
```
Comment (Zheaoli) [inline cmd/nerdctl/pull.go:122]: this function should be removed to pkg package

Comment (yuchanns) [reply cmd/nerdctl/pull.go:122]: Seems that the function is only used by the run command. Maybe better to move to pkg/cmd/run IMO

File: cmd/nerdctl/pull.go (hunk 1 above)
Comment (suyanhanx) [reply cmd/nerdctl/pull.go]: Done

File: cmd/nerdctl/pull.go (hunk 2 above)
Comment (suyanhanx) [reply cmd/nerdctl/pull.go]: Done.

File: cmd/nerdctl/pull.go (hunk 3 above)
Comment (suyanhanx) [reply cmd/nerdctl/pull.go]: Moved to run. Done.

File: cmd/nerdctl/run.go [hunk 4]
```diff
@@ -737,6 +740,79 @@ func createContainer(ctx context.Context, cmd *cobra.Command, client *containerd
 	return container, nil, nil
 }
 
+// TODO This function is copied from /cmd/pull, need to be refactored.
```
Comment (djdongjin) [inline cmd/nerdctl/run.go:743]: what refactor to be needed here? E.g., why can't we import `pkg/cmd/image` and use `image.EnsureImage` directly here?

File: pkg/cmd/image/pull.go [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+	"github.com/containerd/nerdctl/pkg/imgutil"
+	"github.com/containerd/nerdctl/pkg/ipfs"
+	"github.com/containerd/nerdctl/pkg/platformutil"
+	"github.com/containerd/nerdctl/pkg/referenceutil"
+	"github.com/containerd/nerdctl/pkg/strutil"
+	v1 "github.com/opencontainers/image-spec/specs-go/v1"
+	"github.com/sirupsen/logrus"
+)
+
+func Pull(ctx context.Context, stdout io.Writer, stderr io.Writer, rawRef string, options types.PullCommandOptions) error {
```
Comment (djdongjin) [inline pkg/cmd/image/pull.go:40]: I prefer the arguments ordered like this

```golang
Pull(ctx context.Context, rawRef string, options types.PullCommandOptions, stdout io.Writer, stderr io.Writer)
```

`rawRef` is `Pull`'s main target, then `options` decides how we pull, after `pull`, `io.Writer` decides where to write.

Comment (suyanhanx) [reply pkg/cmd/image/pull.go:40]: Make sense.

File: cmd/nerdctl/run.go (hunk 4 above)
Comment (suyanhanx) [reply cmd/nerdctl/run.go]: I have rewritten this function. It should look a bit better.

File: cmd/nerdctl/run.go [hunk 6]
```diff
@@ -737,6 +739,42 @@ func createContainer(ctx context.Context, cmd *cobra.Command, client *containerd
...
+
+	options.GOptions = globalOptions
+
+	ensured, err := image.EnsureImage(ctx, client, rawRef, cmd.OutOrStdout(), cmd.ErrOrStderr(), options, ocispecPlatforms, pull, unpack, quiet)
+	if err != nil {
+		return nil, err
+	}
+
+	return ensured, nil
+}
```
Comment (djdongjin) [inline cmd/nerdctl/run.go:776]: you can remove the helper func `ensureImage`, and call `image.EnsureImage` directly in `generateRootfsOpts`. E.g,: change https://github.com/containerd/nerdctl/blob/2fa1777ceef949786ff1f730c0ad8ef0c7670dc9/cmd/nerdctl/run.go#L763

to

```golang
	options, err := processPullCommandFlagsInRun(cmd)
	if err != nil {
		return nil, err
	}
	options.GOptions = globalOptions
    ensured, err := image.EnsureImage(ctx, client, rawRef, cmd.OutOrStdout(), cmd.ErrOrStderr(), options, ocispecPlatforms, pull, unpack, quiet)
```

File: cmd/nerdctl/run.go [hunk 7]
```diff
@@ -737,6 +739,42 @@ func createContainer(ctx context.Context, cmd *cobra.Command, client *containerd
 	return container, nil, nil
 }
 
+func processPullCommandFlagsInRun(cmd *cobra.Command) (types.PullCommandOptions, error) {
```
Comment (djdongjin) [inline cmd/nerdctl/run.go:742]: Maybe add a comment:

```
// When refactor `nerdctl run`, this func should be removed and replaced by creating a `PullCommandOptions` directly from `RunCommandOptions`
```

File: cmd/nerdctl/run.go (hunk 6 above)
Comment (suyanhanx) [reply cmd/nerdctl/run.go:776]: Make sense. It looks a little more cohesive.

Comment (suyanhanx) [reply cmd/nerdctl/run.go]: Done.

File: cmd/nerdctl/run.go (hunk 7 above)
Comment (suyanhanx) [reply cmd/nerdctl/run.go:743]: Done.

**Review by djdongjin — APPROVED** [review]
LGTM on CI green

**Review by yuchanns — APPROVED** [review]

**Review by Zheaoli — APPROVED** [review]
LGTM on CI green

**Review by ktock — APPROVED** [review]

**Pull request merged**